File: src/ts-packages/sdk/src/dataset-ref.ts
import { ApiClient } from "./api-client.js";
import { Instance } from "./instance.js";
import type { AgentifiedTool, RegisterInput } from "./types.js";

function validateTools(tools: AgentifiedTool[]): void {
  for (const tool of tools) {
    if ("type" in tool && tool.type === "client") {
      throw new Error("Client tools are not yet supported");
    }
    if (!("handler" in tool)) {
      throw new Error(`Tool '${(tool as { name: string }).name}' has no type and no handler`);
    }
  }
}

export class DatasetRef {
  constructor(
    /** @internal */ protected readonly _agentified: { sdk: ApiClient | null; serverUrl: string | null },
    readonly datasetName: string,
  ) {}

  async register(input: RegisterInput): Promise<Instance> {
    validateTools(input.tools);

    const sdk = this._agentified.sdk;
    if (!sdk) throw new Error("Not connected");

    const serverTools = input.tools.map((t) => ({
      name: t.name,
      description: t.description,
      parameters: t.parameters,
      ...("type" in t && t.type === "mcp" ? { type: "mcp", server_uri: t.server } : {}),
    }));
    // Separate ApiClient for registration: it must carry the tool list,
    // while the shared SDK instance is tool-agnostic.
    const regSdk = new ApiClient({
      serverUrl: this._agentified.serverUrl!,
      tools: serverTools,
    });
    await regSdk.register(this.datasetName);

    return new Instance(this.datasetName, this.datasetName, sdk, input.tools);
  }
}

File: src/py-packages/sdk/src/agentified/dataset_ref.py
from __future__ import annotations

from typing import TYPE_CHECKING

from .api_client import ApiClient
from .instance import Instance
from .models import (
    AgentifiedTool,
    ApiClientConfig,
    BackendTool,
    ClientTool,
    McpTool,
    RegisterInput,
    ServerTool,
)

if TYPE_CHECKING:
    from .client import Agentified


class DatasetRef:
    def __init__(self, agentified: Agentified, dataset_name: str) -> None:
        self._agentified = agentified
        self.dataset_name = dataset_name

    async def register(self, input: RegisterInput) -> Instance:
        _validate_tools(input.tools)

        sdk = self._agentified._sdk
        if sdk is None:
            raise RuntimeError("Not connected")

        server_tools = []
        for t in input.tools:
            st = ServerTool(name=t.name, description=t.description, parameters=t.parameters)
            if isinstance(t, McpTool):
                st = ServerTool(
                    name=t.name, description=t.description, parameters=t.parameters,
                    type="mcp", server_uri=t.server,
                )
            elif isinstance(t, BackendTool) and t.always_include:
                st = ServerTool(
                    name=t.name, description=t.description, parameters=t.parameters,
                    always_include=True,
                )
            server_tools.append(st)

        reg_sdk = ApiClient(
            ApiClientConfig(server_url=self._agentified._server_url, tools=server_tools)
        )
        await reg_sdk.register(self.dataset_name)

        return Instance(
            self.dataset_name, self.dataset_name, sdk, input.tools,
            emitter=self._agentified.emitter,
        )


def _validate_tools(tools: list[AgentifiedTool]) -> None:
    for tool in tools:
        if isinstance(tool, ClientTool):
            raise ValueError("Client tools are not yet supported")
        if isinstance(tool, (BackendTool, McpTool)):
            if not callable(tool.handler):
                raise ValueError(f"Tool '{tool.name}' requires a handler")
